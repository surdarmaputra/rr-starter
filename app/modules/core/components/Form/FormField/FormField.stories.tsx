import type { Meta, StoryObj } from '@storybook/react';

import { Input } from '../Input/Input';
import { FormField, type FormFieldProps } from './FormField';

type Story = StoryObj<typeof FormField>;

const meta: Meta = {
  component: FormField,
  title: 'Core/Components/Form/FormField',
  argTypes: {
    hint: {
      control: { type: 'text' },
    },
  },
};

export default meta;

export const Default: Story = {
  args: {
    label: 'Full name',
    isOptional: true,
    errors: [
      'Full name is required',
      'Full name must be at least 2 characters',
    ],
    hint: 'Full name must be at least 2 characters',
  },

  render: (args: FormFieldProps) => (
    <FormField {...args}>
      <Input name="full_name" />
    </FormField>
  ),
};
