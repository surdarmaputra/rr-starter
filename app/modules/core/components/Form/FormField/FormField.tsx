import type { ReactNode } from 'react';

import type { TestableComponentProps } from '~/modules/core/types';

import { Label } from '../Label/Label';

export interface FormFieldProps extends TestableComponentProps {
  children?: ReactNode;
  className?: string;
  errors?: ReactNode[];
  hint?: ReactNode;
  inputId: string;
  isOptional?: boolean;
  label?: string;
}

export function FormField({
  children,
  className,
  'data-testid': dataTestId,
  errors,
  hint,
  inputId,
  isOptional,
  label,
}: FormFieldProps) {
  const labelSecondaryText = isOptional ? '(Optional)' : undefined;
  return (
    <div className={className} data-testid={dataTestId}>
      {Boolean(label) && (
        <Label
          className="mb-2 block font-bold text-letter-title"
          htmlFor={inputId}
          secondaryText={labelSecondaryText}
        >
          {label}
        </Label>
      )}
      {children}
      {Boolean(hint) && (
        <div className="mt-1 text-xs text-letter-disabled">{hint}</div>
      )}
      {Boolean(errors?.length) && (
        <div className="mt-1.5 text-xs text-danger">
          {errors?.map((error, index) => <div key={index}>{error}</div>)}
        </div>
      )}
    </div>
  );
}
