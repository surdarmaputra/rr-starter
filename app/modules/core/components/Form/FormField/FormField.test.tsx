import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, test, vi } from 'vitest';

import { FormField } from './FormField';

describe('FormField', () => {
  function renderComponent(props = {}) {
    return render(
      <FormField
        data-testid="FormField-testid"
        inputId="full-name"
        {...props}
      />,
    );
  }

  test('renders with data-testid', async () => {
    renderComponent();
    await screen.findByTestId('FormField-testid');
  });

  test('renders label with optional text', async () => {
    renderComponent({ label: 'Name', isOptional: true });
    await screen.findByText(/Name/);
    await screen.findByText(/\(Optional\)/);
  });

  test('renders accessible label', async () => {
    const handleChange = vi.fn();
    renderComponent({
      label: 'Name',
      children: <input id="full-name" onChange={handleChange} />,
    });
    const input = await screen.findByLabelText(/Name/);
    fireEvent.change(input, { target: { value: 'New Value' } });

    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(handleChange).toHaveBeenCalledWith(
      expect.objectContaining({
        target: expect.objectContaining({
          value: 'New Value',
        }),
      }),
    );
  });

  test('renders hint', async () => {
    renderComponent({ hint: 'This is a hint' });
    await screen.findByText(/This is a hint/);
  });

  test('renders errors', async () => {
    const errors = ['Error 1', 'Error 2'];
    renderComponent({ errors });
    await screen.findByText(/Error 1/);
    await screen.findByText(/Error 2/);
  });

  test('handles input change', async () => {
    const handleChange = vi.fn();
    const { getByTestId } = renderComponent({
      children: <input data-testid="input-testid" onChange={handleChange} />,
    });

    const input = getByTestId('input-testid');
    fireEvent.change(input, { target: { value: 'New Value' } });

    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(handleChange).toHaveBeenCalledWith(
      expect.objectContaining({
        target: expect.objectContaining({
          value: 'New Value',
        }),
      }),
    );
  });

  test('renders without errors, label, or hint safely', async () => {
    renderComponent();
    const formField = await screen.findByTestId('FormField-testid');
    expect(formField).toMatchSnapshot();
  });

  test('renders with errors, label, and hint safely', async () => {
    renderComponent({
      errors: ['Error 1', 'Error 2'],
      label: 'Name',
      hint: 'This is a hint',
    });
    const formField = await screen.findByTestId('FormField-testid');
    expect(formField).toMatchSnapshot();
  });
});
