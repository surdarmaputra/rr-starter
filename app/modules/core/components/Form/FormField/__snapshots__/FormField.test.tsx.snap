// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormField > renders with errors, label, and hint safely 1`] = `
<div
  data-testid="FormField-testid"
>
  <label
    class="text-sm leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 mb-2 block font-bold text-letter-title"
    for="full_name"
  >
    Name
  </label>
  <div
    class="mt-1 text-xs text-letter-disabled"
  >
    This is a hint
  </div>
  <div
    class="mt-1.5 text-xs text-danger"
  >
    <div>
      Error 1
    </div>
    <div>
      Error 2
    </div>
  </div>
</div>
`;

exports[`FormField > renders without errors, label, or hint safely 1`] = `
<div
  data-testid="FormField-testid"
/>
`;
